types/metadataV14.go

```diff
@@ -0,0 +1,397 @@
+	Version          U8
+	SignedExtensions []SignedExtensionMetadataV14
+}
+
+type SignedExtensionMetadataV14 struct {
+	Identifier       Text
+	Type             Si1LookupTypeID
+	AdditionalSigned Si1LookupTypeID
+}
+
+type PortableRegistryV14 struct {
+	Types []PortableTypeV14
+}
+
+type PalletMetadataV14 struct {
+	Name       Text
+	HasStorage bool
+	Storage    StorageMetadataV14
+	HasCalls   bool
+	Calls      FunctionMetadataV14
+	HasEvents  bool
+	Events     EventMetadataV14
+	Constants  []ConstantMetadataV14
+	HasErrors  bool
+	Errors     ErrorMetadataV14
+	Index      U8
+}
+
+type FunctionMetadataV14 struct {
+	Type Si1LookupTypeID
+}
+
+type EventMetadataV14 struct {
+	Type Si1LookupTypeID
+}
+
+type ConstantMetadataV14 struct {
+	Name  Text
+	Type  Si1LookupTypeID
+	Value Bytes
+	Docs  []Text
+}
+
+type ErrorMetadataV14 struct {
+	Type Si1LookupTypeID
+}
+
+func (m *PalletMetadataV14) Decode(decoder scale.Decoder) error {
+	err := decoder.Decode(&m.Name)
+	if err != nil {
+		return err
+	}
+
+	err = decoder.DecodeOption(&m.HasStorage, &m.Storage)
+	if err != nil {
+		return err
+	}
+
+	err = decoder.DecodeOption(&m.HasCalls, &m.Calls)
+	if err != nil {
+		return err
+	}
+
+	err = decoder.DecodeOption(&m.HasEvents, &m.Events)
+	if err != nil {
+		return err
+	}
+
+	err = decoder.Decode(&m.Constants)
+	if err != nil {
+		return err
+	}
+
+	err = decoder.DecodeOption(&m.HasErrors, &m.Errors)
+	if err != nil {
+		return err
+	}
+
+	return decoder.Decode(&m.Index)
+}
+
+func (m PalletMetadataV14) Encode(encoder scale.Encoder) error {
+	err := encoder.Encode(m.Name)
+	if err != nil {
+		return err
+	}
+
+	err = encoder.EncodeOption(m.HasStorage, m.Storage)
+	if err != nil {
+		return err
+	}
+
+	err = encoder.EncodeOption(m.HasCalls, m.Calls)
+	if err != nil {
+		return err
+	}
+
+	err = encoder.EncodeOption(m.HasEvents, m.Events)
+	if err != nil {
+		return err
+	}
+
+	err = encoder.Encode(m.Constants)
+	if err != nil {
+		return err
+	}
+
+	err = encoder.EncodeOption(m.HasErrors, m.Errors)
+	if err != nil {
+		return err
+	}
+
+	return encoder.Encode(m.Index)
+}
+
+func (m *PalletMetadataV14) FindConstantValue(constant Text) ([]byte, error) {
+	for _, cons := range m.Constants {
+		if cons.Name == constant {
+			return cons.Value, nil
+		}
+	}
+	return nil, fmt.Errorf("could not find constant %s", constant)
+}
+
+type StorageMetadataV14 struct {
+	Prefix Text
+	Items  []StorageEntryMetadataV14
+}
+
+type StorageEntryMetadataV14 struct {
+	Name          Text
+	Modifier      StorageFunctionModifierV0
+	Type          StorageEntryTypeV14
+	Fallback      Bytes
+	Documentation []Text
+}
+
+type MapTypeV14 struct {
+	Hashers []StorageHasherV10
+	Key     Si1LookupTypeID
+	Value   Si1LookupTypeID
+}
+
+func (s StorageEntryMetadataV14) IsPlain() bool {
+	return s.Type.IsPlainType
+}
+
+func (s StorageEntryMetadataV14) IsMap() bool {
+	return s.Type.IsMap
+}
+
+func (s StorageEntryMetadataV14) IsDoubleMap() bool {
+	return false
```

Comment:
so the type is not different for DoubleMap and NMap ?

Reply:
Yeah, so in the v14 there are the Storage entry types: https://github.com/polkadot-js/api/blob/80b581f0df87108c59f71e67d7c5fc5f8c89ec33/packages/types/src/interfaces/metadata/v14.ts#L84-L92

So, `Plain` and `Map`. Therefore I have `IsDoubleMap` returning false. But I just realized that on the other hand I have `IsNMap` returning `s.Type.IsMap`. What do you recon? Should both return false or `s.Type.IsMap`? I'd say the former but let me have your experienced opinion :)

Reply:
ahh cool. So the Map could be Map or double Map or Nmap. We have two approaches here, 
- based on the types, we can find what kind of ma this is and return the bool accordingly. 
- remove these methods completely.

I'm leaning towards the first one since it is straight forward to say what kind of map this is and we also dont break the api.

Reply:
Wait, just read through https://docs.substrate.io/v3/runtime/storage/#:~:text=Storage%20Double%20Map%20%2D%20used%20as,have%20a%20common%20first%20key and my conclusion is that both of these functions should `return false`:

> - **Storage Double Map** - used as an implementation of a storage map with two keys to provide the ability to efficiently remove all entries that have a common first key.
> - **Storage N Map** - used to store a mapping with any arbitrary number of keys, it can be used as a basis to build a Triple Storage Map, a Quadruple Storage Map and so on.

Note how with v14 the Map type `keys` field [became `'key'`](https://github.com/polkadot-js/api/blob/master/packages/types/src/interfaces/metadata/v14.ts#L89)

However, I find it odd that the newer version would drop these advanced Map types, so I may be missing something? 🤔

Reply:
Yes, the key that they are referring to point to the type, which could be single value key like map, tuple with multiple sub keys like double map or nmap. `Map` in storage entry could be any of the three

On the gsrpc api level, these functions should clearly give hints on which kind of map that is. Take a look one storage entry that has StorgeMap and Double map. Take for example - https://github.com/paritytech/substrate/blob/master/frame/staking/src/pallet/mod.rs
It should be clear if you look at how key represent a type which could be tuple or single value key

Reply:
Cool that makes sense 💡  On it 👍

Reply:
Following the discussions in https://github.com/centrifuge/go-substrate-rpc-client/issues/215 and https://github.com/paritytech/frame-metadata/pull/23#issuecomment-970612572, we discontinue `IsDoubleMap` and `IsNMap` and offer `IsMap()` alone, which now abstracts the old 3 variants. As said by the Parity devs, there is no difference between the 3 in their storage representation but only in their Rust signatures, so we follow their advice of following their lead and unifying the 3 concepts.

The only follow up action here is to make it clear for users that from v14, isMap means isMap or IsDouble or IsN and that they probably want to treat the new IsMap as they were treating IsNMap previously. If - for whatever reason we don't foresee - users still need to know the number of keys, they can still do so by checking the length of the Hashers of the map entry.

Reply:
alright. Lets bump the gsrpc version accordingly. Thanks @NunoAlexandre